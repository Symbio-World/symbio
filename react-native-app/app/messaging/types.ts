import { FirebaseMessagingTypes } from '@react-native-firebase/messaging'

// copy paste of FirebaseMessagingTypes.AuthorizationStatus
export enum AuthorizationStatus {
  NOT_DETERMINED = -1,
  DENIED = 0,
  AUTHORIZED = 1,
  PROVISIONAL = 2,
}
export type RemoteMessage = FirebaseMessagingTypes.RemoteMessage
