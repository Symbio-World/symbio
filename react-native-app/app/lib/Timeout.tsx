import * as React from 'react'

type Props = {
  durationInSeconds: number
  onTimeout: () => void
  children: React.ReactElement
}
export const Timeout: React.FC<Props> = ({
  durationInSeconds,
  onTimeout,
  children,
}) => {
  React.useEffect(() => {
    const timer = setTimeout(() => {
      onTimeout()
    }, durationInSeconds)
    return () => clearTimeout(timer)
  }, [onTimeout, durationInSeconds])

  return children
}
