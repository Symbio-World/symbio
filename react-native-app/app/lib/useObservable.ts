import * as React from 'react'
import { Observable } from 'rxjs'

export const useObservable = <Key, Data, Error>(
  key: Key,
  observableCreator: (key: Key) => Observable<Data>,
) => {
  const [data, setData] = React.useState<Data>()
  const [error, setError] = React.useState<Error>()

  React.useEffect(() => {
    const subscription = observableCreator(key).subscribe(setData, setError)
    return () => subscription.unsubscribe()
  }, [key, observableCreator])

  return { data, error }
}
