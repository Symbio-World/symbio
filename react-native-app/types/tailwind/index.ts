export {}

declare module 'react-native-tailwindcss' {
  import {
    TailwindColors as DefaultTailwindColors,
    TailwindStyles as DefaultTailwindStyles,
  } from 'react-native-tailwindcss'

  interface CustomColors {
    primary: any
    recycling: any
    allergies: any
  }

  export interface TailwindStyles extends DefaultTailwindStyles, CustomColors {}
  export interface TailwindColors extends DefaultTailwindColors, CustomColors {}
}
