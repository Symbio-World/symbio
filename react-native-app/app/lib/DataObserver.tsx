import * as React from 'react'
import { Observable } from 'rxjs'
import { ErrorView } from '../ui-kit/ErrorView'
import { Loading } from '../ui-kit/Loading'
import { Timeout } from './Timeout'
import { useObservable } from './useObservable'

type Arg = any
type Data = any
type Error = any
type Props = {
  arg: Arg
  observableCreator: (arg: Arg) => Observable<Data>
  renderSuccess: (data: Data) => React.ReactElement
  timeoutDuration?: number
  renderTimeout?: () => React.ReactElement
  renderError?: (error: Error) => React.ReactElement
}
export const DataObserver: React.FC<Props> = ({
  arg,
  observableCreator,
  renderSuccess,
  timeoutDuration = 5000,
  renderTimeout = () => <ErrorView error="request has timed out" />,
  renderError = (error) => <ErrorView error={error} />,
}) => {
  const { data, error } = useObservable(arg, observableCreator)
  const [hasTimedOut, setHasTimedOut] = React.useState(false)

  const handleTimeout = () => {
    setHasTimedOut(true)
  }

  if (hasTimedOut) return renderTimeout()
  if (error) return renderError(error)
  if (!data) {
    return (
      <Timeout durationInSeconds={timeoutDuration} onTimeout={handleTimeout}>
        <Loading />
      </Timeout>
    )
  }
  return renderSuccess(data)
}
