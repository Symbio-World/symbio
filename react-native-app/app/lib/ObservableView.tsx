import * as React from 'react'
import { Observable } from 'rxjs'
import { ErrorView } from '../ui-kit/ErrorView'
import { TimeoutView } from '../ui-kit/TimeoutView'
import { Loading } from '../ui-kit/Loading'
import { Timeout } from './Timeout'
import { useObservable } from './useObservable'

type Props = {
  observable: Observable<any>
  renderSuccess: (data: any) => React.ReactElement
  timeoutDuration?: number
  renderTimeout?: ({ onRetry }: { onRetry: () => void }) => React.ReactElement
  renderError?: (error: any) => React.ReactElement
}
export const ObservableView: React.FC<Props> = ({
  observable,
  renderSuccess,
  timeoutDuration = 5000,
  renderTimeout = ({ onRetry }) => <TimeoutView onRetry={onRetry} />,
  renderError = (error) => <ErrorView error={error} />,
}) => {
  const { data, error } = useObservable(observable)
  const [hasTimedOut, setHasTimedOut] = React.useState(false)

  const handleTimeout = () => {
    setHasTimedOut(true)
  }

  if (hasTimedOut)
    return renderTimeout({ onRetry: () => setHasTimedOut(false) })
  if (error) return renderError(error)
  // if (!data) {
    return (
      <Timeout durationInSeconds={timeoutDuration} onTimeout={handleTimeout}>
        <Loading />
      </Timeout>
    )
  // }
  return renderSuccess(data)
}
