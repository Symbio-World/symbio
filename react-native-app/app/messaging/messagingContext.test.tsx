import * as React from 'react'
import { Text } from 'react-native'
import { of } from 'rxjs'
import { render, waitForElement } from 'react-native-testing-library'
import {
  createMessagingProvider,
  useMessaging,
  RequestPermission,
  ObserveTokens,
  ObserveMessages,
} from './messagingContext'
import { AuthorizationStatus, RemoteMessage } from './types'
import { ObservableView } from '../lib/ObservableView'

describe('messagingContext', () => {
  let requestPermission: RequestPermission
  let observeTokens: ObserveTokens
  let observeMessages: ObserveMessages

  beforeEach(() => {
    requestPermission = jest.fn(() =>
      Promise.resolve(AuthorizationStatus.AUTHORIZED),
    )
    observeTokens = jest.fn(() => of())
    observeMessages = jest.fn(() => of())
  })

  it('renders correctly', () => {
    const MessagingProvider = createMessagingProvider({
      requestPermission,
      observeTokens,
      observeMessages,
    })
    const { toJSON } = render(
      <MessagingProvider>
        <Text>Test</Text>
      </MessagingProvider>,
    )
    expect(toJSON()).toMatchSnapshot()
  })

  it('requests permission', () => {
    const MessagingProvider = createMessagingProvider({
      requestPermission,
      observeTokens,
      observeMessages,
    })
    render(
      <MessagingProvider>
        <Text>Test</Text>
      </MessagingProvider>,
    )
    expect(requestPermission).toHaveBeenCalled()
  })

  it('calls observeTokens and observeMessages if permission is granted', async () => {
    const MessagingProvider = createMessagingProvider({
      requestPermission,
      observeTokens,
      observeMessages,
    })
    const { getByText } = render(
      <MessagingProvider>
        <Text>Test</Text>
      </MessagingProvider>,
    )

    await waitForElement(() => getByText('Test'))
    expect(observeTokens).toHaveBeenCalled()
    expect(observeMessages).toHaveBeenCalled()
  })

  it('does not call observeTokens and observeMessages if permission is denied', async () => {
    requestPermission = jest.fn(() =>
      Promise.resolve(AuthorizationStatus.DENIED),
    )
    const MessagingProvider = createMessagingProvider({
      requestPermission,
      observeTokens,
      observeMessages,
    })
    const { getByText } = render(
      <MessagingProvider>
        <Text>Test</Text>
      </MessagingProvider>,
    )

    await waitForElement(() => getByText('Test'))
    expect(observeTokens).toHaveBeenCalledTimes(0)
    expect(observeTokens).toHaveBeenCalledTimes(0)
  })

  it('tokens can be consumed', async () => {
    const token = 'token'
    observeTokens = jest.fn(() => of(token))
    const MessagingProvider = createMessagingProvider({
      requestPermission,
      observeTokens,
      observeMessages,
    })
    const Child = () => {
      const { token$ } = useMessaging()

      return (
        <ObservableView
          observable={token$}
          renderSuccess={(token?: string) => <Text>{token}</Text>}
        />
      )
    }
    const { getByText } = render(
      <MessagingProvider>
        <Child />
      </MessagingProvider>,
    )

    await waitForElement(() => getByText(token))
    expect(getByText(token)).toBeDefined()
  })

  it('messages can be consumed', async () => {
    const message = { notification: { body: 'body', title: 'title' } }
    observeMessages = jest.fn(() => of(message))

    const MessagingProvider = createMessagingProvider({
      requestPermission,
      observeTokens,
      observeMessages,
    })
    const Child = () => {
      const { message$ } = useMessaging()

      return (
        <ObservableView
          observable={message$}
          renderSuccess={(m?: RemoteMessage) => (
            <>
              <Text>{m?.notification?.title}</Text>
              <Text>{m?.notification?.body}</Text>
            </>
          )}
        />
      )
    }
    const { getByText } = render(
      <MessagingProvider>
        <Child />
      </MessagingProvider>,
    )

    await waitForElement(() => getByText(message.notification.title))
    expect(getByText(message.notification.title)).toBeDefined()
    expect(getByText(message.notification.body)).toBeDefined()
  })

  // TODO: test unsubscribe
})
