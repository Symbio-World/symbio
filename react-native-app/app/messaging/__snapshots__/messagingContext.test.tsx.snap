// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`messagingContext renders correctly 1`] = `
<Text>
  Test
</Text>
`;
